packages/core/src/external/cda/partition-payload.ts

```diff
@@ -0,0 +1,81 @@
+import { toArray } from "@metriport/shared";
+import { createXMLParser } from "@metriport/shared/common/xml-parser";
+import { sizeInBytes } from "../../util/string";
+import { XMLBuilder } from "fast-xml-parser";
+
+const MAX_CHUNK_SIZE = 10_000_000; // 10MB in bytes
+
+export function partitionPayload(payloadRaw: string): string[] {
+  if (sizeInBytes(payloadRaw) < MAX_CHUNK_SIZE) return [payloadRaw];
+
+  const parser = createXMLParser({
+    ignoreAttributes: false,
+    attributeNamePrefix: "_",
+    removeNSPrefix: true,
+  });
+  const json = parser.parse(payloadRaw);
+
+  if (!json.ClinicalDocument?.component?.structuredBody?.component) {
+    return [payloadRaw];
+  }
+
+  const components = toArray(json.ClinicalDocument?.component?.structuredBody?.component);
+
+  const chunks: string[] = [];
+  let currentComponents: string[] = [];
+  let currentSize = 0;
+
+  const builder = new XMLBuilder({
+    format: false,
+    ignoreAttributes: false,
+    attributeNamePrefix: "_",
+    suppressEmptyNode: true,
+    suppressBooleanAttributes: false,
+  });
+
+  function createChunk(comps: string[]) {
+    const chunk = {
+      ...json,
+      ClinicalDocument: {
+        ...json.ClinicalDocument,
+        component: {
+          ...json.ClinicalDocument.component,
+          structuredBody: {
+            ...json.ClinicalDocument.component.structuredBody,
+            component: comps,
+          },
+        },
+      },
+    };
+    return builder.build(chunk);
+  }
+
+  for (const component of components) {
+    const componentSize = sizeInBytes(JSON.stringify(component));
+
+    if (currentSize + componentSize > MAX_CHUNK_SIZE && currentComponents.length > 0) {
+      chunks.push(createChunk(currentComponents));
+      currentComponents = [];
+      currentSize = 0;
+    }
+
+    if (componentSize > MAX_CHUNK_SIZE) {
+      if (currentComponents.length > 0) {
+        chunks.push(createChunk(currentComponents));
+        currentComponents = [];
+        currentSize = 0;
+      }
+      chunks.push(createChunk([component]));
+      continue;
+    }
```

Comment:
I don't understand the need for these two `if`s - if we enter the second one, we'll enter the first one as well, duplicating the component of the current iteration, right?

Reply:
good catch!
this logic needs to be updated. 

I think the if checks should be as follows:
1. If the currentComp by itself is too large, then:
  - save existing chunk
  - put currentComp into its own chunk
2. If the groupedComps + currentComp are too large, then:
  - add currentComp to groupedComps
  - save them as a new chunk
3. else, just add currentComp to groupedComps

after the for loop, if groupedComps.length > 0, save that chunk

Reply:
I think its simpler than that. Both items (1) and (2) can a single one: IMO item (2) above should use the same logic as item (1), we shouldn't merge the current item w/ the grouped ones it the sum of their sizes is larger than the max.

```ts
  if (groupedSize + currentSize > MAX_CHUNK_SIZE_IN_BYTES) {
    if (groupedComponents.length > 0) {
      chunks.push(createChunk(json, groupedComponents));
    }
    chunks.push(createChunk(json, [currentComponent]));
    // I'd have these here to make sure they're reset, just in case
    groupedComponents = [];
    groupedSize = 0;
    continue;
  }
```

Reply:
Yeah, my code was done w/o testing - by the way, we should have a unit test for this, it would make it much easier to come up with a solid code.

The new version still allows chunks larger than the max, even though the parts are not larger than max.

I think this updated version works, but need testing:

```ts
  let groupedComponents: string[] = [];
  let groupedSize = 0;
  const components = toArray(json.ClinicalDocument?.component?.structuredBody?.component);
  for (const currentComponent of components) {
    const currentSize = sizeInBytes(JSON.stringify(currentComponent));

    if (groupedSize + currentSize > MAX_CHUNK_SIZE_IN_BYTES) {
      if (groupedComponents.length > 0) {
        chunks.push(createChunk(json, groupedComponents));
        groupedComponents = [currentComponent];
        groupedSize = currentSize;
      } else { // first element is larger than max
        chunks.push(createChunk(json, currentComponent));
        groupedComponents = [];
        groupedSize = 0;
      }
      continue;
    }

    groupedComponents.push(currentComponent);
    groupedSize += currentSize;
  }
  if (groupedComponents.length > 0) {
    chunks.push(createChunk(json, groupedComponents));
  }
```